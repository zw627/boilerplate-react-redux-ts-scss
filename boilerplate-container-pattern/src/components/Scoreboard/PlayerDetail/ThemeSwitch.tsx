import React from "react";

import { ThemeSwitchProps } from "Containers/Scoreboard/PlayerDetail/ThemeSwitch";
import toggleBodyClass from "Utils/toggleBodyClass";

const ThemeSwitch: React.FC<ThemeSwitchProps> = ({
  isLightMode,
  toggle,
}: ThemeSwitchProps) => {
  // Theme is switched before mounted, thus no useEffect()
  toggleBodyClass(isLightMode);

  return (
    <button className="mode-switch" onClick={toggle}>
      {isLightMode ? "Light" : "Dark"}
    </button>
  );
};

export default ThemeSwitch;
