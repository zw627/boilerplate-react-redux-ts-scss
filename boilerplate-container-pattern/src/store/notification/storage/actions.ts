// ------------------------------------------------------ //
// Actions

export const SET = "notificationStorage/set";

// ------------------------------------------------------ //
// Action creators

export const set = (visibility: boolean): NotificationStorageActions => ({
  type: "notificationStorage/set",
  payload: {
    visibility,
  },
});

// ------------------------------------------------------ //
// Types - action

interface Set {
  type: typeof SET;
  payload: {
    visibility: boolean;
  };
}

export type NotificationStorageActions = Set | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface NotificationStorageState {
  visibility: boolean;
}
