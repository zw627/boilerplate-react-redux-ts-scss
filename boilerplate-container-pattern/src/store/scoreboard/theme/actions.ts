// ------------------------------------------------------ //
// Actions

export const TOGGLE = "theme/toggle";

// ------------------------------------------------------ //
// Action Creators

export const toggle = (): ThemeActions => ({
  type: "theme/toggle",
});

// ------------------------------------------------------ //
// Types - action

interface Toggle {
  type: typeof TOGGLE;
}

export type ThemeActions = Toggle | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface ThemeState {
  isLightMode: boolean;
}
