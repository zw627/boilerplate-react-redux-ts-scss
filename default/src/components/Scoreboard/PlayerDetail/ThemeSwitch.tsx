import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";

import { AppState } from "Store/index";
import toggleBodyClass from "Utils/toggleBodyClass";
import { isLightModeSelector } from "Utils/selectors";

const ThemeSwitch: React.FC<{}> = () => {
  // State (redux)
  const isLightMode = useSelector((state: AppState) =>
    isLightModeSelector(state)
  );

  // Actions (redux)
  const dispatch = useDispatch();
  const toggle = useCallback(() => dispatch({ type: "theme/toggle" }), [
    dispatch,
  ]);

  // Theme is switched before mounted, thus no useEffect()
  toggleBodyClass(isLightMode);

  return (
    <button className="mode-switch" onClick={toggle}>
      {isLightMode ? "Light" : "Dark"}
    </button>
  );
};

export default React.memo(ThemeSwitch);
