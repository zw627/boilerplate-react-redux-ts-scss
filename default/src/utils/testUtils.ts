// Helper functions for Jest and Enzyme

import { ReactWrapper, ShallowWrapper } from "enzyme";
import { createStore, Store } from "redux";

import { rootReducer } from "Store/index";

// Type for parameters of setupStore()
type setupStorePars = {
  playerList?: {
    id: string;
    name: string;
    score: number;
    created: string;
    updated: string;
    isSelected: boolean;
  }[];
  selectedPlayerId?: string;
  isRunning?: boolean;
  previousTime?: number;
  elapsedTime?: number;
  isLightMode?: boolean;
  visibility?: boolean;
};

// Setup a store with some fake states for testing
export function setupStore({
  playerList = [
    {
      id: "1654A",
      name: "Alpha",
      score: 10,
      created: "11/21/2016",
      updated: "06/12/2018",
      isSelected: false,
    },
    {
      id: "1654B",
      name: "Beta",
      score: 4,
      created: "11/8/2016",
      updated: "11/9/2017",
      isSelected: false,
    },
  ],
  isRunning = false,
  previousTime = 0,
  elapsedTime = 0,
  isLightMode = false,
  visibility = false,
}: setupStorePars = {}): Store {
  return createStore(rootReducer, {
    notification: {
      storage: {
        visibility,
      },
    },
    scoreboard: {
      player: {
        playerList,
      },
      stopwatch: {
        isRunning,
        previousTime,
        elapsedTime,
      },
      theme: {
        isLightMode,
      },
    },
  });
}

// Multiple toEqual, useful when testing a table
export function multiToEqual([...args]: Array<string[] | number[]>): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    expect(arg[0]).toEqual(arg[1]);
  });
}

export function matchLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [element, number]: [string, number]
): void {
  expect(wrapper.find(element).length).toEqual(number);
}

// Multiple matchLength
export function matchMultiLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [...args]: [string, number][]
): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    matchLength(wrapper, [arg[0], arg[1]]);
  });
}

// export const matchText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [element, index]: [string, number],
//   text: string
// ): void => {
//   expect(
//     wrapper
//       .find(element)
//       .at(index)
//       .text()
//   ).toEqual(text);
// };

// export const matchMultiText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [...args]
// ): void => {
//   if ([...args].length > 0)
//     [...args].map(arg => {
//       matchText(wrapper, [arg[0], arg[1]], arg[2]);
//     });
// };
