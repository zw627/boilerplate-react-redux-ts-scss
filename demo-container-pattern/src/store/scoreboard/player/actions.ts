// ------------------------------------------------------ //
// Actions

export const ADD = "player/add";
export const REMOVE = "player/remove";
export const UPDATE_SCORE = "player/update";
export const SELECT = "player/select";

// ------------------------------------------------------ //
// Action creators

export const add = (id: string, name: string, date: string): PlayerActions => ({
  type: "player/add",
  payload: {
    id,
    name,
    date,
  },
});

export const remove = (id: string): PlayerActions => ({
  type: "player/remove",
  payload: {
    id,
  },
});

export const update = (
  id: string,
  delta: number,
  date: string
): PlayerActions => ({
  type: "player/update",
  payload: {
    id,
    delta,
    date,
  },
});

export const select = (id: string): PlayerActions => ({
  type: "player/select",
  payload: {
    id,
  },
});

// ------------------------------------------------------ //
// Types - action

interface Add {
  type: typeof ADD;
  payload: {
    id: string;
    name: string;
    date: string;
  };
}

interface Remove {
  type: typeof REMOVE;
  payload: {
    id: string;
  };
}

interface Update {
  type: typeof UPDATE_SCORE;
  payload: {
    id: string;
    delta: number;
    date: string;
  };
}

interface Select {
  type: typeof SELECT;
  payload: {
    id: string;
  };
}

export type PlayerActions =
  | Add
  | Remove
  | Update
  | Select
  | { type: undefined };

// ------------------------------------------------------ //
// Types - state

export interface PlayerObject {
  id: string;
  name: string;
  score: number;
  created: string;
  updated: string;
  isSelected: boolean;
}

export interface PlayerState {
  playerList: PlayerObject[];
}
