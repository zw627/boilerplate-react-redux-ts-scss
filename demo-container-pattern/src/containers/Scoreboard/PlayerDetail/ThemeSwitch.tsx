import { connect } from "react-redux";

// Component
import ThemeSwitch from "Components/Scoreboard/PlayerDetail/ThemeSwitch";

// Store
import { AppState } from "Store/index";
import { isLightModeSelector } from "Utils/selectors";
import { toggle, ThemeState } from "Store/scoreboard/theme/actions";

export interface ThemeSwitchProps extends ThemeState {
  toggle: typeof toggle;
}

const mapStateToProps = (state: AppState): ThemeState => ({
  isLightMode: isLightModeSelector(state),
});

const actionCreators = {
  toggle,
};

export default connect(mapStateToProps, actionCreators)(ThemeSwitch);
