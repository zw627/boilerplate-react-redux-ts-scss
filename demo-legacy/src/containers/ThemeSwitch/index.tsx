import { connect } from "react-redux";

// Component
import ThemeSwitch from "Components/ThemeSwtich";

// Store
import { AppState } from "Store/index";
import { isLightModeSelector } from "Store/selectors";
import { toggle, ThemeState } from "Store/theme/actions";

export interface ThemeSwitchProps extends ThemeState {
  toggle: typeof toggle;
}

const mapStateToProps = (state: AppState): ThemeState => ({
  isLightMode: isLightModeSelector(state),
});

const actionCreators = {
  toggle,
};

export default connect(mapStateToProps, actionCreators)(ThemeSwitch);
