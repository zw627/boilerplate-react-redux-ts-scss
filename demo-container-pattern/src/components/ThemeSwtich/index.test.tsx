import React from "react";
import { Provider } from "react-redux";
import { Store } from "redux";
import { mount, ReactWrapper } from "enzyme";

import { toggle } from "Store/theme/actions";
import ThemeSwitch from "Components/ThemeSwtich";
import { setupStore } from "Utils/testUtils";
import { isLightModeSelector } from "Store/selectors";

import { AppState } from "Store/index";
import { ThemeSwitchProps } from "Containers/ThemeSwitch";

jest.mock("Store/theme/actions");

// Pass down the props from mock store to wrapper
function setupProps(
  state = setupStore().getState() as AppState
): ThemeSwitchProps {
  return {
    isLightMode: isLightModeSelector(state),
    toggle,
  };
}

function setupWrapper(
  mockStore = setupStore(),
  props = setupProps(mockStore.getState())
): ReactWrapper {
  return mount(
    <Provider store={mockStore}>
      <ThemeSwitch {...props} />
    </Provider>
  ) as ReactWrapper;
}

describe("Scoreboard/ThemeSwitch", () => {
  let mockStore: Store;
  let wrapper: ReactWrapper;

  // Setup wrapper
  beforeEach(() => {
    wrapper = setupWrapper();
  });

  test("compare to the last snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should display the slider on the left side if light mode", () => {
    mockStore = setupStore({ isLightMode: true });
    wrapper = setupWrapper(mockStore);
    expect(wrapper.find(".theme-switch-slider.left").length).toEqual(1);
  });

  it("should display the slider on the right side if dark mode", () => {
    expect(wrapper.find(".theme-switch-slider.right").length).toEqual(1);
  });

  it("should handle switchTheme() dispatch", () => {
    wrapper.find("button").simulate("click");
    expect(toggle).toHaveBeenCalledTimes(1);
  });
});
