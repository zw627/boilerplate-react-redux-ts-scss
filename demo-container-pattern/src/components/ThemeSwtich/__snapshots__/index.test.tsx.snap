// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard/ThemeSwitch compare to the last snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ThemeSwitch
    isLightMode={false}
    toggle={[MockFunction]}
  >
    <button
      className="theme-switch"
      onClick={[MockFunction]}
    >
      <img
        alt="Sun"
        className="theme-switch-icon-sun"
        src={Object {}}
      />
      <img
        alt="Moon"
        className="theme-switch-icon-moon"
        src={Object {}}
      />
      <span
        className="theme-switch-slider right"
      />
    </button>
  </ThemeSwitch>
</Provider>
`;
